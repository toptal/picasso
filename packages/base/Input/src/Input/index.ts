import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props as InternalInputProps } from './Input'
export { default as Input } from './Input'
export type InputProps = OmitInternalProps<InternalInputProps>
/** @deprecated [FX-4714] Use InputProps instead */
export type Props = InputProps
