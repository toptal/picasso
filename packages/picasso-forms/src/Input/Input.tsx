import React, { useEffect, useMemo } from 'react'
import type { InputProps } from '@toptal/picasso'
import { Input as PicassoInput } from '@toptal/picasso'
import { useForm } from 'react-final-form'

import type { FieldProps } from '../Field'
import type { Props as FieldLabelProps } from '../FieldLabel'
import FieldLabel from '../FieldLabel'
import InputField from '../InputField'

export type FormInputProps = Omit<InputProps, 'onResetClick'> & {
  /** Callback invoked when reset button was clicked */
  onResetClick?: (set: (value: string) => void) => void
}
export type Props = FormInputProps &
  FieldProps<InputProps['value']> &
  FieldLabelProps

const warnAutocompleteDisabledInput = (name?: string) => {
  const autocompleteDisabled =
    name && /^(((field|input)(_|-)?\d+)|tan|otp|title|captcha)$/.test(name)

  if (autocompleteDisabled) {
    console.warn(`
In Chromium-based browsers, autocomplete might be disabled for input[name="${name}"].
Known bug: https://bugs.chromium.org/p/chromium/issues/detail?id=1255609
    `)
  }
}

export const Input = React.forwardRef<HTMLInputElement, Props>((props, ref) => {
  useEffect(() => {
    // TODO: remove this console.warn completely when Chrome fixes autocomplete issue
    // Regex is taken from https://source.chromium.org/chromium/chromium/src/+/main:components/autofill/core/browser/autocomplete_history_manager.cc;l=53;drc=1d2260f9ed19c755db1631b7fb9b1ba216b323dc
    // according to https://bugs.chromium.org/p/chromium/issues/detail?id=1255609
    warnAutocompleteDisabledInput(props.name)
  }, [props.name])

  const {
    mutators: { setHasMultilineCounter },
  } = useForm()

  const { label, labelEndAdornment, titleCase, ...rest } = props
  const { multiline, rows, rowsMax } = props

  const alignment = useMemo(() => {
    if (!multiline) {
      return 'middle'
    }

    if (Number(rows ?? 1) >= 2 || Number(rowsMax ?? 1) >= 2) {
      return 'top'
    }

    return 'middle'
  }, [multiline, rows, rowsMax])

  return (
    <InputField<FormInputProps>
      {...rest}
      setHasMultilineCounter={setHasMultilineCounter}
      label={
        label ? (
          <FieldLabel
            name={props.name}
            required={props.required}
            label={label}
            labelEndAdornment={labelEndAdornment}
            titleCase={titleCase}
            alignment={alignment}
          />
        ) : null
      }
    >
      {(inputProps: InputProps) => <PicassoInput {...inputProps} ref={ref} />}
    </InputField>
  )
})

Input.defaultProps = {}

Input.displayName = 'Input'

export default Input
