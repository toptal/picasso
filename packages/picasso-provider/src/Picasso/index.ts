export { default } from './Picasso'
export { default as PicassoLight } from './PicassoLight'
export { default as FontsLoader } from './FontsLoader'
export { default as FixViewport } from './FixViewport'
export {
  default as NotificationsProvider,
  NotificationsProviderProps,
} from './NotificationsProvider'
