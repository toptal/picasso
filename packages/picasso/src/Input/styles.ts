import type { Theme } from '@material-ui/core/styles'
import { createStyles } from '@material-ui/core/styles'

import highlightAutofillStyles from '../InputBase/highlightStyles'
import '../InputBase/styles'
import '../InputLabel/styles'
import '../OutlinedInput/styles'
import '../InputAdornment/styles'

export default (theme: Theme) =>
  createStyles({
    root: {
      fontSize: '1rem',
      backgroundColor: theme.palette.common.white,
      cursor: 'text',
    },
    rootMultiline: {
      height: 'auto',
    },
    inputMultilineResizable: {
      resize: 'vertical',
    },
    ...highlightAutofillStyles(theme),
  })
