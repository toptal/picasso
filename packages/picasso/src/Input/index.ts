import { OmitInternalProps } from '@toptal/picasso-shared'

import { Props } from './Input'
export { default } from './Input'
export type Props = OmitInternalProps<Props>
