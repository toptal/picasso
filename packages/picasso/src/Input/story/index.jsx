import { Input } from '../Input'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Forms').createPage(
  'Input',
  `Input fields are UI elements through which users submit information to the system.
    Input fields should be clearly labeled by the topic to ensure users know exactly what is being asked of them.`
)

page
  .createChapter()
  .addComponentDocs({
    component: Input,
    name: 'Input',
    additionalDocs: {
      iconPosition: {
        type: 'enum',
        enums: ['start', 'end'],
      },
    },
  })
  .addExample('Input/story/Default.example.tsx', {
    title: 'Default',
    effect: async (testPage, makeScreenshot) => {
      await testPage.hover('[data-testid="input"]')
      await makeScreenshot()

      await testPage.click('[data-testid="input"]')
      await makeScreenshot()
    },
  })
  .addExample('Input/story/Disabled.example.tsx', 'Disabled')
  .addExample('Input/story/Status.example.tsx', {
    title: 'Status',
    effect: async (testPage, makeScreenshot) => {
      await testPage.hover('[data-testid="input"]')
      await makeScreenshot()

      await testPage.click('[data-testid="input"]')
      await makeScreenshot()
    },
    takeScreenshot: false,
  })
  .addExample('Input/story/WithIcon.example.tsx', 'With icon')
  .addExample('Input/story/Sizes.example.tsx', 'Sizes')
  .addExample('Input/story/FullWidth.example.tsx', 'Full width')
  .addExample('Input/story/Multiline.example.tsx', 'Multiline | Textarea')
  .addExample(
    'Input/story/MultilineExpand.example.tsx',
    'Multiline expand and resize | Textarea'
  )
  .addExample(
    'Input/story/AutoFill.example.tsx',
    'AutoComplete defined as email'
  )
  .addExample('Input/story/WithLimit.example.tsx', 'With Limit')
  .addExample('Input/story/ResetButton.example.tsx', {
    title: 'With reset button',
    takeScreenshot: false,
  })
  .addExample('Input/story/Refs.example.tsx', {
    title: 'Refs',
    takeScreenshot: false,
  })
