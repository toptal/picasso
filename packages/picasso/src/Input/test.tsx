import React from 'react'
import { render } from '@testing-library/react'
import SearchIcon from '@material-ui/icons/Search'
import Picasso from '@toptal/picasso-shared'

import Input from './Input'

test('renders icon at the end', () => {
  const { container } = render(
    <Picasso loadFonts={false}>
      <Input icon={<SearchIcon />} />
    </Picasso>
  )

  expect(container).toMatchSnapshot()
})

test('renders icon at the beginning', () => {
  const { container } = render(
    <Picasso loadFonts={false}>
      <Input icon={<SearchIcon />} iconPosition='start' />
    </Picasso>
  )

  expect(container).toMatchSnapshot()
})

test('shows counter for regular input', () => {
  const { container } = render(
    <Picasso loadFonts={false}>
      <Input limit={10} />
    </Picasso>
  )

  expect(container).toMatchSnapshot()
})

test('shows counter for multiline input', () => {
  const { container } = render(
    <Picasso loadFonts={false}>
      <Input multiline rows={4} limit={10} />
    </Picasso>
  )

  expect(container).toMatchSnapshot()
})

test('is focused when autoFocus', () => {
  const { getByPlaceholderText } = render(
    <Picasso loadFonts={false}>
      <Input autoFocus placeholder='test input' />
    </Picasso>
  )

  const input = getByPlaceholderText('test input')

  expect(document.activeElement).toEqual(input)
})
