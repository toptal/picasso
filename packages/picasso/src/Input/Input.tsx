/* eslint-disable max-lines */
import React, {
  ReactNode,
  ChangeEvent,
  InputHTMLAttributes,
  MouseEvent,
  forwardRef,
} from 'react'
import cx from 'classnames'
import { Theme, makeStyles } from '@material-ui/core/styles'
import {
  BaseProps,
  SizeType,
  useHasMultilineCounter,
} from '@toptal/picasso-shared'

import OutlinedInput, { BaseInputProps, Status } from '../OutlinedInput'
import { disableUnsupportedProps } from '../utils'
import { FeatureOptions } from '../utils/disable-unsupported-props'
import styles from './styles'
import { usePropDeprecationWarning } from '../utils/use-deprecation-warnings'
import InputLimitAdornment, {
  InputLimitAdornmentProps,
} from '../InputLimitAdornment'
import InputIconAdornment, {
  InputIconAdornmentProps,
} from '../InputIconAdornment'
import Container from '../Container'
import InputValidIconAdornment from '../InputValidIconAdornment'

export interface Props
  extends BaseProps,
    Omit<InputHTMLAttributes<HTMLInputElement>, 'size'> {
  /** The id of the `input` element. */
  id?: string
  /** Name attribute of the input element */
  name?: string
  /** The default `input` element value. Use when the component is not controlled. */
  defaultValue?: string
  /** The value of the `input` element, required for a controlled component. */
  value?: string
  /** Placeholder for value */
  placeholder?: string
  /**
   * @deprecated Use the `status` prop instead to both support success and error states
   * Indicate whether `Input` is in error state
   */
  error?: boolean
  /** Indicate `Input` status */
  status?: Status
  /** If true, the `Input` will be disabled */
  disabled?: boolean
  /** Width of the component */
  width?: 'full' | 'shrink' | 'auto'
  /** Whether icon should be placed at the beginning or end of the `Input` */
  iconPosition?: InputIconAdornmentProps['position']
  /** Specify icon which should be rendered inside Input */
  icon?: InputIconAdornmentProps['icon']
  inputProps?: BaseInputProps
  /** Whether `Input` should be rendered as `TextArea` or not */
  multiline?: boolean
  /** Whether a multiline can be manually resized by the user, requires multiline prop to be set to true */
  multilineResizable?: boolean
  /** If true, the input element will be focused during the first mount */
  autoFocus?: boolean
  /** Specify rows amount for `TextArea` */
  rows?: string | number
  /** Maximum number of rows to display when multiline option is set to true. */
  rowsMax?: string | number
  /** Type attribute of the Input element. It should be a valid HTML5 input type */
  type?: string
  /** Adds element at the start of the input - can't be used in combination with `iconPosition: start` */
  startAdornment?: ReactNode
  /** Adds element at the end of the input - can't be used in combination with `iconPosition: end` */
  endAdornment?: ReactNode
  /**  Callback invoked when `Input` changes its state */
  onChange?: (
    event: ChangeEvent<
      HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement
    >
  ) => void
  /** Adds a counter of characters (ignored in combination with `counter: entered`) */
  limit?: InputLimitAdornmentProps['limit']
  /** Type of the counter of characters */
  counter?: InputLimitAdornmentProps['counter']
  /** Component size */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** Whether to render reset icon when there is a value in the input */
  enableReset?: boolean
  /** Callback invoked when reset button was clicked */
  onResetClick?: (
    event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>
  ) => void
  /** Ref of the input outline */
  outlineRef?: React.Ref<HTMLElement>
  testIds?: {
    inputAdornment?: string
    resetButton?: string
    validIcon?: string
  }
  setHasMultilineCounter?: (name: string, hasCounter: boolean) => void
}

type StartAdornmentProps = Pick<Props, 'icon' | 'iconPosition' | 'disabled'>

type EndAdornmentProps = Pick<
  Props,
  | 'icon'
  | 'iconPosition'
  | 'disabled'
  | 'multiline'
  | 'limit'
  | 'counter'
  | 'status'
  | 'testIds'
> & { charsLength?: number; showCounter: boolean }

const useStyles = makeStyles<Theme>(styles, { name: 'PicassoInput' })

const hasRemainingCounter = ({
  counter,
  limit,
}: Pick<Props, 'counter' | 'limit'>) =>
  Boolean(counter === 'remaining' && limit)

const hasEnteredCounter = ({ counter }: Pick<Props, 'counter'>) =>
  counter === 'entered'

const hasCounter = ({ counter, limit }: Pick<Props, 'counter' | 'limit'>) =>
  hasRemainingCounter({ counter, limit }) || hasEnteredCounter({ counter })

const hasMultilineAdornment = ({
  multiline,
  status,
  counter,
  limit,
}: Pick<Props, 'multiline' | 'status' | 'counter' | 'limit'>) =>
  multiline && (status === 'success' || hasCounter({ counter, limit }))

const StartAdornment = ({
  icon,
  iconPosition,
  disabled,
}: StartAdornmentProps) => {
  if (!icon || iconPosition !== 'start') {
    return null
  }

  return <InputIconAdornment disabled={disabled} position='start' icon={icon} />
}

const EndAdornment = (props: EndAdornmentProps) => {
  const {
    icon,
    iconPosition,
    disabled,
    limit,
    multiline,
    charsLength,
    testIds,
    counter,
    showCounter,
  } = props

  if (icon && iconPosition === 'end') {
    return <InputIconAdornment disabled={disabled} position='end' icon={icon} />
  }

  if (!multiline && showCounter) {
    return (
      <InputLimitAdornment
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        charsLength={charsLength!}
        multiline={multiline}
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        counter={counter!}
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        limit={limit!}
        testIds={testIds}
      />
    )
  }

  return null
}

type MultilineAdornmentProps = Pick<
  Props,
  'counter' | 'limit' | 'status' | 'testIds'
> &
  Pick<EndAdornmentProps, 'charsLength' | 'showCounter'>

const MultilineAdornment = ({
  charsLength,
  counter,
  limit,
  showCounter,
  status,
  testIds,
}: MultilineAdornmentProps) => {
  return (
    <Container>
      {showCounter && (
        <InputLimitAdornment
          testIds={testIds}
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          charsLength={charsLength!}
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          counter={counter!}
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          limit={limit!}
          multiline
        />
      )}
      {status === 'success' && (
        <InputValidIconAdornment data-testid={testIds?.validIcon} multiline />
      )}
    </Container>
  )
}

const purifyProps = (props: Props) => {
  const sizeOptions: FeatureOptions<Props> = {
    featureProps: {
      size: 'small',
    },
    unsupportedProps: {
      multiline: false,
      icon: undefined,
      startAdornment: undefined,
      endAdornment: undefined,
      limit: undefined,
    },
  }

  return disableUnsupportedProps('Input', props, sizeOptions)
}

export const Input = forwardRef<HTMLInputElement, Props>(function Input(
  props,
  ref
) {
  const {
    id,
    name,
    defaultValue,
    value,
    placeholder,
    error,
    status,
    disabled,
    icon,
    iconPosition,
    inputProps,
    children,
    multiline,
    multilineResizable,
    autoFocus,
    width,
    className,
    style,
    rows,
    rowsMax,
    type,
    onChange,
    onClick,
    startAdornment,
    endAdornment,
    limit,
    counter,
    size,
    enableReset,
    onResetClick,
    outlineRef,
    testIds,
    setHasMultilineCounter,
    ...rest
  } = purifyProps(props)

  usePropDeprecationWarning({
    props,
    name: 'error',
    componentName: 'Input',
    description:
      'Use the `status` prop instead. `error` is deprecated and will be removed in the next major release.',
  })

  const charsLength = value ? value.length : 0

  const classes = useStyles()

  const showCounter = !!charsLength && hasCounter({ counter, limit })

  useHasMultilineCounter(name, showCounter && multiline, setHasMultilineCounter)

  return (
    <>
      <OutlinedInput
        ref={outlineRef}
        inputRef={ref}
        className={className}
        style={style}
        classes={{
          root: cx(classes.root, {
            [classes.rootMultiline]: multiline,
          }),
          input: cx(classes.input, {
            [classes.inputMultilineResizable]: multiline && multilineResizable,
          }),
        }}
        id={id}
        name={name}
        defaultValue={defaultValue}
        value={value}
        placeholder={placeholder}
        status={error ? 'error' : status}
        disabled={disabled}
        multiline={multiline}
        autoFocus={autoFocus}
        rows={rows}
        rowsMax={rowsMax}
        type={type}
        width={width}
        size={size}
        onClick={onClick}
        // html attributes
        inputProps={{
          ...rest,
          ...inputProps,
        }}
        startAdornment={
          startAdornment || (
            <StartAdornment
              icon={icon}
              iconPosition={iconPosition}
              disabled={disabled}
            />
          )
        }
        endAdornment={
          endAdornment || (
            <EndAdornment
              icon={icon}
              iconPosition={iconPosition}
              disabled={disabled}
              limit={limit}
              charsLength={charsLength}
              multiline={multiline}
              counter={counter}
              status={status}
              testIds={testIds}
              showCounter={showCounter}
            />
          )
        }
        onChange={onChange}
        enableReset={enableReset}
        onResetClick={onResetClick}
        testIds={testIds}
      >
        {children}
      </OutlinedInput>
      {hasMultilineAdornment({
        multiline,
        status,
        limit,
        counter,
      }) && (
        <MultilineAdornment
          charsLength={charsLength}
          status={status}
          testIds={testIds}
          showCounter={showCounter}
          counter={counter}
          limit={limit}
        />
      )}
    </>
  )
})

Input.defaultProps = {
  autoComplete: 'none',
  counter: 'remaining',
  iconPosition: 'start',
  multiline: false,
  size: 'medium',
  width: 'auto',
  onChange: () => {},
  onResetClick: () => {},
  status: 'default',
}

Input.displayName = 'Input'

export default Input
