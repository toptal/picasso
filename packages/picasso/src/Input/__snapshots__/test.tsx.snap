// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders icon in the beginning 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root PicassoOutlinedInput-root PicassoInput-root PicassoOutlinedInput-rootAuto PicassoOutlinedInput-rootMedium MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
    >
      <div
        class="MuiInputAdornment-root PicassoInputAdornment-root MuiInputAdornment-disablePointerEvents MuiInputAdornment-positionStart"
      >
        <svg
          class="PicassoSvgSearch16-root PicassoInputIconAdornment-icon"
          style="min-width: 16px; min-height: 16px;"
          viewBox="0 0 16 16"
        >
          <path
            d="m10.582 9.874 4.625 4.626-.707.707-4.626-4.625a6 6 0 1 1 .707-.707ZM6 11A5 5 0 1 0 6 1a5 5 0 0 0 0 10Z"
          />
        </svg>
      </div>
      <input
        aria-invalid="false"
        autocomplete="none"
        class="MuiInputBase-input MuiOutlinedInput-input PicassoOutlinedInput-input PicassoOutlinedInput-inputMedium MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline PicassoOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Input renders icon in the end 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root PicassoOutlinedInput-root PicassoInput-root PicassoOutlinedInput-rootAuto PicassoOutlinedInput-rootMedium MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
    >
      <div
        class="MuiInputAdornment-root PicassoInputAdornment-root MuiInputAdornment-disablePointerEvents MuiInputAdornment-positionStart"
      >
        <svg
          class="PicassoSvgSearch16-root PicassoInputIconAdornment-icon"
          style="min-width: 16px; min-height: 16px;"
          viewBox="0 0 16 16"
        >
          <path
            d="m10.582 9.874 4.625 4.626-.707.707-4.626-4.625a6 6 0 1 1 .707-.707ZM6 11A5 5 0 1 0 6 1a5 5 0 0 0 0 10Z"
          />
        </svg>
      </div>
      <input
        aria-invalid="false"
        autocomplete="none"
        class="MuiInputBase-input MuiOutlinedInput-input PicassoOutlinedInput-input PicassoOutlinedInput-inputMedium MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline PicassoOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Input should show manual resize handler 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root PicassoOutlinedInput-root PicassoInput-root PicassoInput-rootMultiline PicassoOutlinedInput-rootAuto PicassoOutlinedInput-rootMedium MuiInputBase-multiline MuiOutlinedInput-multiline MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
    >
      <textarea
        aria-invalid="false"
        autocomplete="none"
        class="MuiInputBase-input MuiOutlinedInput-input PicassoOutlinedInput-input PicassoInput-inputMultilineResizable PicassoOutlinedInput-inputMedium MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline PicassoOutlinedInput-inputMultiline MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        rows="4"
      >
        Some value
      </textarea>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline PicassoOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Input should show reset button 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root PicassoOutlinedInput-root PicassoInput-root PicassoOutlinedInput-rootAuto PicassoOutlinedInput-rootMedium MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
    >
      <input
        aria-invalid="false"
        autocomplete="none"
        class="MuiInputBase-input MuiOutlinedInput-input PicassoOutlinedInput-input PicassoOutlinedInput-inputMedium MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        type="text"
        value="Some value"
      />
      <div
        class="MuiInputAdornment-root PicassoInputAdornment-root PicassoOutlinedInput-resetButton PicassoOutlinedInput-resetButtonDirty MuiInputAdornment-positionEnd"
        data-testid="reset-adornment"
      >
        <button
          class="MuiButtonBase-root PicassoButton-small PicassoButton-primary PicassoButton-root PicassoButtonCircular-flat PicassoButtonCircular-root"
          data-component-type="button"
          role="reset"
          tabindex="-1"
          type="button"
        >
          <span
            class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoContainer-inline PicassoButton-content"
          >
            <svg
              class="PicassoSvgCloseMinor16-root PicassoButton-icon"
              style="min-width: 16px; min-height: 16px;"
              viewBox="0 0 16 16"
            >
              <path
                d="m8.707 8 3.5 3.5-.707.707-3.5-3.5-3.5 3.5-.707-.707 3.5-3.5-3.5-3.5.707-.707 3.5 3.5 3.5-3.5.707.707-3.5 3.5Z"
              />
            </svg>
          </span>
        </button>
      </div>
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline PicassoOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Input shows counter for multiline input 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root PicassoOutlinedInput-root PicassoInput-root PicassoInput-rootMultiline PicassoInput-rootMultilineLimiter PicassoOutlinedInput-rootAuto PicassoOutlinedInput-rootMedium MuiInputBase-multiline MuiOutlinedInput-multiline MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
    >
      <textarea
        aria-invalid="false"
        autocomplete="none"
        class="MuiInputBase-input MuiOutlinedInput-input PicassoOutlinedInput-input PicassoOutlinedInput-inputMedium MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline PicassoOutlinedInput-inputMultiline MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        rows="4"
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline PicassoOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Input shows counter for regular input 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root PicassoOutlinedInput-root PicassoInput-root PicassoOutlinedInput-rootAuto PicassoOutlinedInput-rootMedium MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
    >
      <input
        aria-invalid="false"
        autocomplete="none"
        class="MuiInputBase-input MuiOutlinedInput-input PicassoOutlinedInput-input PicassoOutlinedInput-inputMedium MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline PicassoOutlinedInput-notchedOutline"
        style="padding-left: 8px;"
      >
        <legend
          class="PrivateNotchedOutline-legend"
          style="width: 0.01px;"
        >
          <span>
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`Input shows excess chars for multiline input with exceeded limit 1`] = `
<div
  class="MuiInputAdornment-root PicassoInputAdornment-root PicassoInputLimitAdornment-limiterMultiline MuiInputAdornment-disablePointerEvents MuiInputAdornment-positionEnd"
  data-testid="limit-adornment-multiline-label"
>
  <span
    class="PicassoInputLimitAdornment-limiterLabel PicassoInputLimitAdornment-limiterLabelError"
  >
    1
     
    over the limit
  </span>
</div>
`;

exports[`Input shows remaining chars for for multiline input with limit 1`] = `
<div
  class="MuiInputAdornment-root PicassoInputAdornment-root PicassoInputLimitAdornment-limiterMultiline MuiInputAdornment-disablePointerEvents MuiInputAdornment-positionEnd"
  data-testid="limit-adornment-multiline-label"
>
  <span
    class="PicassoInputLimitAdornment-limiterLabel PicassoInputLimitAdornment-limiterLabelError"
  >
    0
     
    characters left
  </span>
</div>
`;
